drivers/adc/ad7799/iio_ad7799.c

```diff
@@ -0,0 +1,125 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include "iio_types.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "ad7799.h"
+#include "util.h"
+#include "error.h"
+
+/**
+ * @brief Read ADC Register.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+ssize_t ad7799_iio_reg_read(void *device, char *buf, size_t len,
+			    const struct iio_ch_info *channel)
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+	uint32_t data;
+
+	uint32_t addr = srt_to_uint32(buf);
+
+	ad7799_read(dev, addr, &data);
+
+	return snprintf(buf, len, "%d", data);
+}
+
+/**
+ * @brief Read ADC Channel data.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+ssize_t ad7799_iio_channel_read(void *device, char *buf, size_t len,
+				const struct iio_ch_info *channel)
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+	uint32_t data;
+
+	ad7799_get_channel(dev, channel->ch_num, &data);
+
+	return snprintf(buf, len, "%d", data);
+}
+
+/**
+ * @brief Read ADC Gain value.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+ssize_t ad7799_iio_gain_read(void *device, char *buf, size_t len,
+			     const struct iio_ch_info *channel)
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+	uint8_t gain;
+
+	ad7799_get_gain(dev, &gain);
+
+	return snprintf(buf, len, "%d", gain);
+}
+
+/**
+ * @brief Write ADC Gain value.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+ssize_t ad7799_iio_gain_write(void *device, char *buf, size_t len,
+			      const struct iio_ch_info *channel)
+{
+	struct ad7799_dev *dev = (struct ad7799_dev *)device;
+
+	uint32_t gain = srt_to_uint32(buf);
+
+	ad7799_set_gain(dev, (uint8_t)gain);
+
+	return len;
```

Comment:
looks like there should be some more error handling throughout the file.
what if the driver functions return failure ?